---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```
# grass

<!-- badges: start -->
[![Travis build status](https://travis-ci.com/jackmwolf/grass_alpha.svg?branch=master)](https://travis-ci.com/jackmwolf/grass_alpha)
<!-- badges: end -->

`grass` (Genetic Regression Approximation through Summary Statistics) is an in-development R package to describe various regression models using only pre-computed summary statistics (PCSS) from genome-wide association studies (GWASs) and PCSS repositories such as [GeneAtlas](http://geneatlas.roslin.ed.ac.uk/).
This eliminates the logistic, privacy, and access concerns that accompany the use of individual patient-level data (IPD).


The following figure highlights the information typically needed to perform regression analysis on a set of $m$ phenotypes with $p$ covariates when IPD is available, and the PCSS that are commonly needed to approximate this same model in `grass`.

![Data needed for analysis using IPD compared to that when using PCSS](./man/figures/IPDvsPCSS.png)

Currently, `grass` supports the linear modeling of complex phenotypes defined via functions of other phenotypes.
Supported functions include:

* linear combinations (e.g. $\phi_1y_1 + \phi_2y_2$)
* products (e.g. $y_1\circ y_2$)
* logical combinations (e.g. $y_1\wedge y_2$ or $y_1\vee y_2$)

## Installation

grass is not currently available on CRAN.

You can install the in-development version of grass from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("jackmwolf/grass")
```
## Examples

We will walk through two examples using grass to model combinations of phenotypes using PCSS and then compare our results to those found using IPD.

```{r}
library(grass)
```


### Principal Component Analysis

Let's model the first principal component score of three phenotypes using PCSS.

First, we'll load in some data. We have a SNP's minor allele counts (`g`), a continuous covariate (`x`), and three continuous phenotypes (`y1`, `y2`, and `y3`).

```{r}
dat <- grass::cont_data[c("g", "x", "y1", "y2", "y3")]
head(dat)
```

First, we need our assumed summary statistics. 
We are careful so that the order of `means` and `covs` are the same, and that their last elements are the phenotypes of interest.

```{r}
means <- colMeans(dat)
covs  <- cov(dat)
n     <- nrow(dat)
```

In addition,  need our weights. 
These are the the first principal component vector of the phenotype covariance matrix, and they are in the same order as the final elements of `means` and `covs`.

```{r}
SigmaY <- covs[c("y1", "y2", "y3"), c("y1", "y2", "y3")]
phi <- eigen(SigmaY)$vectors[, 1]
```

Then, we can calculate the linear model by using `calculate_lm_combo()`.

```{r}
model_pcss <- calculate_lm_combo(
  means = means, covs = covs, n = n, phi = phi, add_intercept = TRUE
  )
model_pcss
```

Here's the same model using individual patient data. 

```{r}
pc_1 <- prcomp(x = dat[c("y1", "y2", "y3")])$x[, "PC1"]

mod_ipd <- lm(pc_1 ~ 1 + g + x, data = dat)
summary(mod_ipd)$coef
summary(mod_ipd)$sigma^2
```

In this case, our coefficient estimates for `g` and `x` are off by a factor of -1; this is because we picked the opposite vector of principal component weights to `prcomp`.
This distinction in sign is arbitrary (see the note in `?prcomp`).

### Logical Combination

In this example we will approximate a linear model where our response is the logical combination "$y_1$ or $y_1$" ($y_1\vee y_2$).

First we need data with binary phenotypes.

```{r}
dat <- grass::bin_data[c("g", "x", "y1", "y2")]
head(dat)
```

Once again we will organize our PCSS such that `means` and `covs` have the same order with the phenotypes at the end.

```{r}
means <- colMeans(dat)
covs  <- cov(dat)
n     <- nrow(dat)
```

We also need to describe the distributions of both of our predictors through objects of class `predictor`. 
(See `?new_predictor`.)
grass has shortcut functions to create `predictor` objects for common types of variables, which we will use to create a list of `predictor`s.
(Note that the order `predictors` matches the order of the predictors in `means` and `covs`.)

```{r}
predictors <- list(
  g = new_predictor_snp(maf = means["g"] / 2),
  x = new_predictor_normal(mean = means["x"], sd = sqrt(covs["x", "x"]))
)
class(predictors[[1]])
```

Then we can approximate the linear model using `approx_or()`.

```{r}
model_pcss <- approx_or(
  means = means, covs = covs, n = n, predictors = predictors, add_intercept = TRUE
)
model_pcss
```

And here's the result we would get using IPD:

```{r}
model_ipd <- lm(y1 | y2 ~ 1 + g + x, data = dat)
summary(model_ipd)$coef
summary(model_ipd)$sigma^2
```

## Future Work

* Incorporate support for function notation (E.g. `y1 * y2 ~ 1 + g + x`) instead of depending on the order of input PCSS

* Print model output in a more similar format to `summary.lm`

## References
Following are the key references for the functions in this package

* Wolf, J.M., Barnard, M., Xueting, X., Ryder, N., Westra, J., and Tintle, N. 
  (2020). Computationally efficient, exact, covariate-adjusted genetic principal
  component analysis by leveraging individual marker summary statistics from 
  large biobanks. *Pacific Symposium on Biocomputing*, 25, 719-730. 
  [https://doi.org/10.1142/9789811215636_0063](https://doi.org/10.1142/9789811215636_0063).
  
* Gasdaska A., Friend D., Chen R., Westra J., Zawistowski M., Lindsey W. and 
  Tintle N. (2019) Leveraging summary statistics to make inferences about 
  complex phenotypes in large biobanks. *Pacific Symposium on Biocomputing*, 24, 
  391-402.
  [https://doi.org/10.1142/9789813279827_0036](https://doi.org/10.1142/9789813279827_0036).
